(BKR-648) Refactored to simplify puppet-agent installation

kevpl opened the pull request.

kevpl commented:
/cc @kylog @joshcooper @puppetlabs/beaker

kylog commented on lib/beaker/dsl/install_utils/foss_utils.rb:
```diff
@@ -331,24 +331,14 @@ def install_puppet_agent_on(hosts, opts)
-              else
-                host.install_package('puppet-agent')
-              end
+            when /el-|fedora|sles/
```
centos too?

kylog commented on lib/beaker/dsl/install_utils/foss_utils.rb:
```diff
@@ -859,12 +854,13 @@ def install_puppetlabs_release_repo_on( hosts, repo = nil, opts = options )
 
             case variant
-            when /^(fedora|el|centos)$/
+            when /^(fedora|el|centos|sles)$/
```
In puppet-agent components (puppet, etc), we moved to matching against `el-` rather than `el` because the latter matches `elcapitan`.

kevpl replied:
should this be an issue for the variant?  This should still be `osx` in this case, while `elcapitan` should be the codename:

    variant, version, arch, codename = host['platform'].to_array

When I hit about on my mac, I see "OS X El Capitan", which leads me to believe that this shouldn't be an issue for the variant.  `elcapitan` wouldn't pass our platform string validation as a platform name, are there team's using it as such?

kevpl replied:
we do have times when we're using this against `host['platform']` in general, however, so I'll fix those

kevpl replied:
ok, so I did an audit of the code related to this PR for this problem.  Taking a second look at what you were saying, I see that uses of the variant and `el-` would be incorrect, as would uses of `host['platform']` and `el` (the case that you brought up).

Weirdly, I did have these issues, but they came up exclusively in `unix/file.rb`.  I've fixed them, and am pushing up the changes momentarily. Thanks @kylog!

kylog commented on lib/beaker/dsl/install_utils/foss_utils.rb:
```diff
@@ -904,12 +920,19 @@ def install_puppetlabs_release_repo_on( hosts, repo = nil, opts = options )
                                   repo_configs_dir = 'tmp/repo_configs',
                                   opts = options )
           variant, version, arch, codename = host['platform'].to_array
+          if variant !~ /^(fedora|el|centos|debian|ubuntu|cumulus)$/
```
does this need sles too?

kevpl replied:
if you checkout `install_puppet_agent_dev_repo_on` in foss_utils.rb:1061-74, you'll see that the sles logic wasn't refactored to call this method, but to keep its old logic.  This was because there were no artifacts for me to test possibly going down this path that I found (info on this in my JIRA ticket comments).

kylog replied:
Okay. Tbh I'm fuzzy on what the logic should be for the PE-only agent platforms (like `sles`), so I was asking out of curiosity at this point.

colinPL commented on lib/beaker/dsl/install_utils/foss_utils.rb:
```diff
@@ -918,110 +941,44 @@ def install_puppetlabs_dev_repo ( host, package_name, build_version,
           # url type
           _, protocol, hostname = opts[:dev_builds_url].partition /.*:\/\//
           dev_builds_url = protocol + hostname
+          dev_builds_url = opts[:dev_builds_url] if variant =~ /^(fedora|el|centos)$/
```
Should this be overwriting `dev_builds_url` or appending to it?

kevpl replied:
overwriting, as this is something that was in the case statement before I refactored it out. I've refactored it here so that everyone can just use `dev_builds_url`, rather than either `dev_builds_url` or `opts[:dev_builds_url]`, as they had been before

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-beaker.delivery.puppetlabs.net//job/qe_beaker_btc-intn/2025/
Test PASSed.

kevpl commented:
![](https://s-media-cache-ak0.pinimg.com/originals/97/54/1c/97541ce16475852992666587aaa90162.jpg)

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-beaker.delivery.puppetlabs.net//job/qe_beaker_btc-intn/2026/
Test FAILed.

kevpl commented:
I killed this job manually, because I've got new changes to push

colinPL commented on lib/beaker/host/mac.rb:
```diff
@@ -23,6 +23,50 @@ def platform_defaults
+      end
+      if not link_exists?("#{opts[:download_url]}/#{release_path_end}/#{release_file}") # oops, try the old stuff
+        # the old school
+        release_path_end.chomp!(path_chunk) #remove chunk that didn't work
```
`release_path_end` is equal to `path_chunk`. This line is most likely not needed and the line below can be `=` instead of `<<`.

colinPL commented on lib/beaker/host/unix/file.rb:
```diff
@@ -29,4 +29,196 @@ def file_exist?(path)
+  # @param [String] repo_name Name of the repo to check for the package in
+  #
+  # @return [String, nil] Path to the package repo if it exists, nil otherwise
+  def repo_path_exists?(buildserver_url, package_name, build_version, repo_name)
```
The `?` suggests this returns a boolean, but instead it returns `[String, nil]`. This should be renamed, something like `get_repo_path`.

colinPL commented on lib/beaker/host/windows.rb:
```diff
@@ -43,6 +43,28 @@ def determine_ssh_server
+    #   time. Note that it will not fail if not provided, however
+    #
+    # @return [String, String] Path to the directory and filename of the package, respectively
+    def puppet_agent_dev_package_info( puppet_collection = nil, puppet_agent_version = nil, opts = {} )
```
Should `puppet_collection` or `puppet_agent_version` be used anywhere below?

kevpl replied:
no.  This is a method override, so it needs to match the signature of the others, even though Windows doesn't require that information for its use.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-beaker.delivery.puppetlabs.net//job/qe_beaker_btc-intn/2027/
Test PASSed.

colinPL commented:
Assuming this passes CI, :+1:

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-beaker.delivery.puppetlabs.net//job/qe_beaker_btc-intn/2030/
Test PASSed.

colinPL commented:
@kylog - can you give a final review?

joshcooper commented:
@kevpl is the "single method to install puppet agent" `install_puppet_agent_on` or `install_puppet_agent_dev_repo_on`?

joshcooper commented on lib/beaker/dsl/install_utils/foss_utils.rb:
```diff
@@ -1204,57 +1084,21 @@ def install_puppet_agent_dev_repo_on( hosts, opts )
               install_msi_on(host, onhost_copied_file, {}, opts)
             when /^osx$/
-              host.install_package("#{mac_pkg_name}*")
+              host.install_package("puppet-agent-#{opts[:puppet_agent_version]}")
```
This isn't working for OSX, it appears to be missing the osversion:

```
Install repositories on target machines...
  Fetching: http://builds.delivery.puppetlabs.net/puppet-agent/b9b44eab32118af58c4d8bd97d0e57451f32534e/repos/apple/10.9/PC1/x86_64/puppet-agent-1.3.2.164.gb9b44ea-1.osx10.9.dmg
    and saving to tmp/repo_configs/osx/puppet-agent-1.3.2.164.gb9b44ea-1.osx10.9.dmg
  localhost $ scp tmp/repo_configs/osx/puppet-agent-1.3.2.164.gb9b44ea-1.osx10.9.dmg osx109-64-1:/var/root {:ignore => }

  gr9xks5oysxviqm.delivery.puppetlabs.net (osx109-64-1) 13:56:29$ hdiutil attach puppet-agent-1.3.2.164.gb9b44ea.dmg
    hdiutil: attach failed - No such file or directory
```

joshcooper commented:
I wrote a little script to ensure beaker could install puppet-agent on all currently supported platforms using hiera. It's missing some of the netdev platforms, so that'd be good to test also before merging this.

```bash
#!/bin/bash

set -e

test_targets=( centos4-64a centos4-32a debian7-64a debian7-32a fedora21-64a fedora21-32a fedora22-64a fedora22-32a debian6-64a debian6-32a debian7-64a debian7-32a debian8-64a debian8-32a osx109-64a osx1010-64a osx1011-64a redhat5-64a redhat5-32a redhat6-64a redhat6-32a redhat7-64a sles10-64a sles10-32a sles11-64a sles11-32a sles12-64a solaris10-64a solaris11-64a ubuntu1004-64a ubuntu1004-32a ubuntu1204-64a ubuntu1204-32a ubuntu1404-64a ubuntu1404-32a ubuntu1504-64a ubuntu1504-32a windows2008r2-64a windows2008r2-6432a windows2012r2-64a windows2012r2-6432a )

for i in "${test_targets[@]}"
do
    echo "Testing $i"
    bundle exec rake ci:test:aio TEST_TARGET=$i SUITE_VERSION=1.3.2.164.gb9b44ea SHA=b9b44eab32118af58c4d8bd97d0e57451f32534e TESTS=tests/yaml_backend/00-setup.rb
done
```

You can test this PR against hiera:

```
$ cd ~/work/beaker
$ git checkout pr/1044
$ cd ~/work/hiera
$ git checkout stable
$ git pull upstream stable
$ export BEAKER_VERSION=file:///Users/josh/work/beaker
$ cd acceptance
$ bundle install --path .bundle
```

Then edit setup/aio/pre-suite/010_Install.rb so it calls the appropriate beaker method to install puppet-agent, e.g.

```ruby
step "Install puppet-agent..." do
  opts = {
    :puppet_collection    => 'PC1',
    :puppet_agent_sha     => ENV['SHA'],
    :puppet_agent_version => ENV['SUITE_VERSION'] || ENV['SHA']
  }

  install_puppet_agent_on(hosts, opts)
end
```

Then execute the bash script.

kevpl commented:
@joshcooper the one method depends on whether you want a released or dev build of puppet-agent:

- dev build: `install_puppet_agent_dev_repo_on`
- released: `install_puppet_agent_on`

Doing this allows us to keep the old signatures and parameter structure for old users of these methods.

If we want to have one method that arbitrates between these two, it'll have to be an additional one, but looking at the facter pre-suites, I thought having `...dev_repo_on` being the main method to install from builds.pl.n would work.  If I was wrong about that, and having this additional method is necessary for you, just let me know, and I can add it either here or in a new ticket, whichever works best for you.

joshcooper commented:
By "released" do you mean it's been pushed to {yum|apt|downloads}.pl.com? If so, "dev" builds are fine so long as we're able to install both SHA and tagged builds, e.g. puppet-agent-1.3.2. I'll take a look at updating hiera's acceptance to use `install_puppet_agent_dev_repo_on` for all hosts, and verify we can install on all supported platforms.

kevpl commented:
yes, that's what I mean.  Thanks for kicking off jobs to put the expanded artifact folder up!  I'll increase my testing load as well.

The latest push was to fix the OSX issue that you brought up before.  There was an asterisk at the end of the old code that didn't make it through the refactor, which broke `host.install_package`, which relied on the globbing functionality to find the right name & resolve paths.  It's a little unfortunate, but I can see how it's more resilient than knowing OSX pathing details, even though those ones appear consistent between 10.9 & 10.11.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-beaker.delivery.puppetlabs.net//job/qe_beaker_btc-intn/2034/
Test FAILed.

joshcooper commented:
I pushed a [hiera branch](https://github.com/joshcooper/hiera/tree/ticket/stable/HI-486-simplify-acceptance-setup) that calls [`install_puppet_agent_dev_repo_on`](https://github.com/joshcooper/hiera/blob/ticket/stable/HI-486-simplify-acceptance-setup/acceptance/setup/aio/pre-suite/010_Install.rb#L7-L15), and I executed hiera acceptance using beaker from this PR. I executed acceptance on one platform of each flavor, and see failures for platforms that curl the repo_config and try to install through yum, apt, etc. See https://gist.github.com/joshcooper/1ff15bcab961b9732b4e.

For example, on centos4, I see it curls the http://builds.delivery.puppetlabs.net/puppet-agent/b9b44eab32118af58c4d8bd97d0e57451f32534e/repo_configs/rpm/pl-puppet-agent-b9b44eab32118af58c4d8bd97d0e57451f32534e-el-4-x86_64.repo and that is a valid URL. The downloaded file looks correct:

```
[root@cwapgpdubsim3d2 yum.repos.d]# cat pl-puppet-agent-b9b44eab32118af58c4d8bd97d0e57451f32534e-el-4-x86_64.repo
[pl-puppet-agent-b9b44eab32118af58c4d8bd97d0e57451f32534e]
name=PL Repo for puppet-agent at commit b9b44eab32118af58c4d8bd97d0e57451f32534e
baseurl=http://builds.puppetlabs.lan/puppet-agent/b9b44eab32118af58c4d8bd97d0e57451f32534e/repos/el/4/PC1/x86_64/
enabled=1
gpgcheck=0
```

and baseurl is a valid URL. If I manually install puppet, it works ok:

```
[root@cwapgpdubsim3d2 yum.repos.d]# yum install puppet-agent
Setting up Install Process
Setting up repositories
pl-puppet-agent-b9b44eab3 100% |=========================| 2.4 kB    00:00
update                    100% |=========================|  951 B    00:00
base                      100% |=========================| 1.1 kB    00:00
addons                    100% |=========================|  951 B    00:00
extras                    100% |=========================| 1.1 kB    00:00
Reading repository metadata in from local files
0de25c0c34129d6d6b96358df 100% |=========================| 1.8 kB    00:00
pl-puppet-: ################################################## 1/1
Parsing package install arguments
Resolving Dependencies
--> Populating transaction set with selected packages. Please wait.
---> Downloading header for puppet-agent to pack into transaction set.
puppet-agent-1.3.2.164.gb 100% |=========================| 522 kB    00:00
---> Package puppet-agent.x86_64 0:1.3.2.164.gb9b44ea-1.el4 set to be updated
--> Running transaction check

Dependencies Resolved

=============================================================================
 Package                 Arch       Version          Repository        Size
=============================================================================
Installing:
 puppet-agent            x86_64     1.3.2.164.gb9b44ea-1.el4  pl-puppet-agent-b9b44eab32118af58c4d8bd97d0e57451f32534e   35 M

Transaction Summary
=============================================================================
Install      1 Package(s)
Update       0 Package(s)
Remove       0 Package(s)
Total download size: 35 M
Is this ok [y/N]: y
Downloading Packages:
(1/1): puppet-agent-1.3.2 100% |=========================|  35 MB    00:00
Running Transaction Test
Finished Transaction Test
Transaction Test Succeeded
Running Transaction
  Installing: puppet-agent                 ######################### [1/1]

Installed: puppet-agent.x86_64 0:1.3.2.164.gb9b44ea-1.el4
Complete!
```

kevpl commented:
@puppetlabs-jenkins retest this please

kevpl commented:
@joshcooper I'm running facter acceptance from your scripts in the same way, and I'm seeing successful installs on the platforms you mentioned (not on centos4, however, because [of this issue](https://github.com/puppetlabs/beaker/blob/master/lib/beaker/host/unix/pkg.rb#L69-L70)).

Due to the output that you're seeing, I think that the failures are caused by trying to download the repo itself.  The quick fix for this is to provide `dev_builds_url => PC1` in the options hash.  This seems like it could be a problem, if not at least a discussion, so I'll provide details for that discussion in the next paragraph:

You can see the code for this (in the PR) in [unix/file.rb:157](https://github.com/puppetlabs/beaker/pull/1044/files#diff-236029029ebe0d6af5a423fcb6e9c35bR157).  `dev_build_repos` gets the full array of repos for a package.  If you checkout where the parameter _build_repos_ comes from, you'll see it gets used in [foss_utils.rb:959](https://github.com/puppetlabs/beaker/pull/1044/files#diff-b0b73189db06b67008309aabcc69a743R959), passing in `opts[:dev_builds_repos]`. These are actually from the global options hash, because `install_puppet_agent_dev_repos_on` doesn't pass its opts hash to `install_puppetlabs_dev_repo`.  I could see reason for changing [foss_utils.rb:1063](https://github.com/puppetlabs/beaker/pull/1044/files#diff-b0b73189db06b67008309aabcc69a743R1063) to this:

    opts[:dev_builds_repos] ||= [ opts[:puppet_collection] ]
    install_puppetlabs_dev_repo( host, 'puppet-agent', sha, nil, opts )

A few footnotes about that change (mostly to myself):
- I'd have to make the `install_puppetlabs_dev_repo` _repo_configs_dir_ parameter default to nil with a line like this, so the caller doesn't have to know its defaults (unless you have a better idea of how I should do this):
    ```
    repo_configs_dir ||= 'tmp/repo_configs'
    ```
- we have had issues before where we used to default an options hash to the global one, and when we pulled that out we hit errors.  In this case, I don't think that's a problem because we don't pass the opts hash to any lower methods, and the only two used are built there `:dev_builds_url` ([foss_utils.rb:937](https://github.com/puppetlabs/beaker/pull/1044/files#diff-b0b73189db06b67008309aabcc69a743R937)) and needed from the outside either way `:dev_builds_repos`.

joshcooper commented:
@kevpl facter acceptance only calls `install_puppet_agent_dev_repos` for non-rpm/deb platforms: https://github.com/puppetlabs/facter/blob/stable/acceptance/lib/puppet/acceptance/install_utils.rb#L165.

kevpl commented:
I have a changed local copy that just calls `install_puppet_agent_dev_repo_on` for `hosts`:

    step "Installing from install_puppet_agent_dev_repo_on" do
      sha = ENV['SHA'] || ENV['GIT_SHA']
      opts = {
        :puppet_collection => 'PC1',
        :puppet_agent_sha => ENV['SHA'],
        :puppet_agent_version => ENV['SUITE_VERSION'] || ENV['SHA']
      }
      @logger.warn( "dev_builds_repos before: #{options[:dev_builds_repos]}")
      options[:dev_builds_repos] ||= []
      options[:dev_builds_repos] << 'PC1'
      @logger.warn( "dev_builds_repos after: #{options[:dev_builds_repos]}")
      install_puppet_agent_dev_repo_on( hosts, opts )
    end

joshcooper commented:
So adding `:dev_builds_repos => ['PC1']` "works" in that beaker installs puppet-agent, but it's curling the repo config and the puppet-agent rpm/deb, and installing the puppet-agent rpm/deb package directly, instead of using `yum install -y puppet-agent`:

```
* Install puppet-agent...
  Fetching: http://builds.delivery.puppetlabs.net/puppet-agent/b9b44eab32118af58c4d8bd97d0e57451f32534e/repos/el/7/PC1/x86_64/puppet-agent-1.3.2.164.gb9b44ea-1.el7.x86_64.rpm
    and saving to tmp/repo_configs/el/puppet-agent-1.3.2.164.gb9b44ea-1.el7.x86_64.rpm
  localhost $ scp tmp/repo_configs/el/puppet-agent-1.3.2.164.gb9b44ea-1.el7.x86_64.rpm redhat7-64-1:/root {:ignore => }

  s12lyizjj44null.delivery.puppetlabs.net (redhat7-64-1) 14:00:36$ rpm -ivh /root/puppet-agent-1.3.2.164.gb9b44ea-1.el7.x86_64.rpm
```

If you're having issues installing centos4, it sounds like you're not using your beaker PR, since that is only an issue in beaker#master (or beaker gem)

joshcooper commented on lib/beaker/dsl/install_utils/foss_utils.rb:
```diff
@@ -1097,98 +1054,21 @@ def install_puppet_agent_dev_repo_on( hosts, opts )
-                opts[:puppet_collection], opts[:puppet_agent_version] )
+            when /^(fedora|el|centos|debian|ubuntu|cumulus)$/
+              sha = opts[:puppet_agent_sha] || opts[:puppet_agent_version]
+              install_puppetlabs_dev_repo( host, 'puppet-agent', sha)
```
need to pass `opts` to the method, otherwise `install_puppetlabs_dev_repo` will default to global options.

joshcooper replied:
This seems to do the trick:

```ruby
install_puppetlabs_dev_repo( host, 'puppet-agent', sha, opts[:copy_base_local], opts )
```

Using `copy_base_local` instead of `copy_dir_local`, because using the latter causes `variant` to be added twice, e.g. `tmp/repo_configs/debian/debian/wheezy`

joshcooper replied:
Ah, I see that's what you were talking about earlier.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-beaker.delivery.puppetlabs.net//job/qe_beaker_btc-intn/2036/
Test FAILed.

kevpl commented:
cancelled manually.  Before that, I added the code that fixes the `:dev_builds_repos` issue.

@joshcooper, I think the last thing left that you brought up was installing the rpm file. The code before this wasn't installing the package files, but it was still copying them over to the host.  I've removed the copying, so now this should be behaving as you expect.  I'll push this up in a sec, just doing some last testing.  Let me know if you're seeing anymore issues in your testing, or if we're good.

joshcooper commented:
@kevpl getting closer! I'm able to install both rpm & deb with and without specifying `:dev_builds_repos`, but I noticed a few issues:

debian7:

* we're creating the [/root/puppet-agent directory](https://gist.github.com/joshcooper/1e527e4bd81b5842d55e#file-gistfile1-txt-L3) but AFAIK never using it
* we're looking for a [link at /root/puppet-agent/wheezy/PC1, failing to find it, and falling back to `main`](https://gist.github.com/joshcooper/1e527e4bd81b5842d55e#file-gistfile1-txt-L10-L15). Since we never copy anything to that location, we can probably skip looking in /root/puppet-agent, or default to `main` to start with?
* we try to [rewrite the URL](https://gist.github.com/joshcooper/1e527e4bd81b5842d55e#file-gistfile1-txt-L17) in `/etc/apt/sources.list.d/*.list` to `file:///root/puppet-agent/wheezy` but the sed command isn't matching, so the file continues to point to http://builds.delivery.puppetlabs.net (which is actually good). The sed doesn't match because the *list file contains `builds.puppetlabs.lan` but the regex is looking for `builds.delivery.puppetlabs.net` and also because there's an extra backslash in `\s\+` (should be `\s+`). 
```
# cat /etc/apt/sources.list.d/pl-puppet-agent-3e3aa2bf08096133d8ba82269dea5b17454ca288-wheezy.list
# Packages for puppet-agent built from ref 3e3aa2bf08096133d8ba82269dea5b17454ca288
deb http://builds.puppetlabs.lan/puppet-agent/3e3aa2bf08096133d8ba82269dea5b17454ca288/repos/apt/wheezy wheezy PC1
```
So I'm thinking the sed command business isn't necessary and can be removed?
* We're executing `apt-get update` twice, once in https://gist.github.com/joshcooper/1e527e4bd81b5842d55e#file-gistfile1-txt-L21 and again in https://gist.github.com/joshcooper/1e527e4bd81b5842d55e#file-gistfile1-txt-L58

redhat7:

* we're creating the [/root/puppet-agent directory](https://gist.github.com/joshcooper/31ff01ae977a7e65ae54#file-gistfile1-txt-L3) but not using it
* we try to [rewrite the URL](https://gist.github.com/joshcooper/31ff01ae977a7e65ae54#file-gistfile1-txt-L9) in `/etc/yum.repos.d/*.repo` but it doesn't match due to the `builds.puppetlabs.lan`, and could be skipped entirely:

```
# cat /etc/yum.repos.d/pl-puppet-agent-3e3aa2bf08096133d8ba82269dea5b17454ca288-el-7-x86_64.repo
[pl-puppet-agent-3e3aa2bf08096133d8ba82269dea5b17454ca288]
name=PL Repo for puppet-agent at commit 3e3aa2bf08096133d8ba82269dea5b17454ca288
baseurl=http://builds.puppetlabs.lan/puppet-agent/3e3aa2bf08096133d8ba82269dea5b17454ca288/repos/el/7/PC1/x86_64/
enabled=1
gpgcheck=0
```

The following commit seems to work for both deb7 and redhat7: https://github.com/joshcooper/beaker/commit/436760f3303cc879bdf7713a580da32a7579de15, but in doing so I noticed we're only setting AIO defaults in the debian case? Maybe that needs to be moved out of the case statement?

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-beaker.delivery.puppetlabs.net//job/qe_beaker_btc-intn/2037/
Test PASSed.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-beaker.delivery.puppetlabs.net//job/qe_beaker_btc-intn/2039/
Test FAILed.

kevpl commented:
@puppetlabs-jenkins retest this please

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-beaker.delivery.puppetlabs.net//job/qe_beaker_btc-intn/2040/
Test PASSed.

joshcooper commented on lib/beaker/host/unix/file.rb:
```diff
@@ -29,4 +29,201 @@ def file_exist?(path)
+    end
+
+    repo_path
+  end
```
It looks like `repo_path_exists`, `dev_build_repos` and `repo_path` are dead code since we're no longer copying the puppet-agent repo to the local SUT and rewriting the repo_config to point to the local filesystem.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-beaker.delivery.puppetlabs.net//job/qe_beaker_btc-intn/2043/
Test PASSed.

kevpl commented:
Thanks for catching that @joshcooper!  What do you think now?

joshcooper commented:
@kevpl looks good, I'm just running through all of the currently supported platforms

joshcooper commented:
ubuntu-15.10 has problems, it appears it's missing a call to `apt-get update`, and is unable to install `puppet-agent`:

```
$ bundle exec rake ci:test:aio SUITE_VERSION=1.3.2.193.gda2ae6d SHA=da2ae6dfcef8ad1c36cbdf63e8c854ef5f06804a TEST_TARGET=ubuntu1504-64a
...
* Install puppet-agent...
  Fetching: http://builds.delivery.puppetlabs.net/puppet-agent/da2ae6dfcef8ad1c36cbdf63e8c854ef5f06804a/repo_configs/deb/pl-puppet-agent-da2ae6dfcef8ad1c36cbdf63e8c854ef5f06804a-wily.list
    and saving to tmp/repo_configs/ubuntu/pl-puppet-agent-da2ae6dfcef8ad1c36cbdf63e8c854ef5f06804a-wily.list
  localhost $ scp tmp/repo_configs/ubuntu/pl-puppet-agent-da2ae6dfcef8ad1c36cbdf63e8c854ef5f06804a-wily.list ubuntu1510-64-1:/etc/apt/sources.list.d {:ignore => }

  sj2xu43snjgh0s1.delivery.puppetlabs.net (ubuntu1510-64-1) 15:31:14$ echo "/opt/puppetlabs/bin"
    /opt/puppetlabs/bin

  sj2xu43snjgh0s1.delivery.puppetlabs.net (ubuntu1510-64-1) executed in 0.01 seconds

  sj2xu43snjgh0s1.delivery.puppetlabs.net (ubuntu1510-64-1) 15:31:14$ grep ^PATH=.*\/opt\/puppetlabs\/bin ~/.ssh/environment
    PATH=PATH:/opt/puppetlabs/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games

  sj2xu43snjgh0s1.delivery.puppetlabs.net (ubuntu1510-64-1) executed in 0.01 seconds

  sj2xu43snjgh0s1.delivery.puppetlabs.net (ubuntu1510-64-1) 15:31:14$ grep ^PATH=.*PATH ~/.ssh/environment
    PATH=PATH:/opt/puppetlabs/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games

  sj2xu43snjgh0s1.delivery.puppetlabs.net (ubuntu1510-64-1) executed in 0.01 seconds

  sj2xu43snjgh0s1.delivery.puppetlabs.net (ubuntu1510-64-1) 15:31:14$ apt-get install --force-yes  -y puppet-agent
    Reading package lists...
    Building dependency tree...
    Reading state information...
    E    :     Unable to locate package puppet-agent
```

However 15.04 installs ok:

```
$ bundle exec rake ci:test:aio SUITE_VERSION=1.3.2.193.gda2ae6d SHA=da2ae6dfcef8ad1c36cbdf63e8c854ef5f06804a TEST_TARGET=ubuntu1504-64a
...
* Install puppet-agent...
  Fetching: http://builds.delivery.puppetlabs.net/puppet-agent/da2ae6dfcef8ad1c36cbdf63e8c854ef5f06804a/repo_configs/deb/pl-puppet-agent-da2ae6dfcef8ad1c36cbdf63e8c854ef5f06804a-vivid.list
    and saving to tmp/repo_configs/ubuntu/pl-puppet-agent-da2ae6dfcef8ad1c36cbdf63e8c854ef5f06804a-vivid.list
  localhost $ scp tmp/repo_configs/ubuntu/pl-puppet-agent-da2ae6dfcef8ad1c36cbdf63e8c854ef5f06804a-vivid.list ubuntu1504-64-1:/etc/apt/sources.list.d {:ignore => }

  i239onwbhyvo67w.delivery.puppetlabs.net (ubuntu1504-64-1) 15:33:57$ echo "/opt/puppetlabs/bin"
    /opt/puppetlabs/bin

  i239onwbhyvo67w.delivery.puppetlabs.net (ubuntu1504-64-1) executed in 0.01 seconds

  i239onwbhyvo67w.delivery.puppetlabs.net (ubuntu1504-64-1) 15:33:57$ grep ^PATH=.*\/opt\/puppetlabs\/bin ~/.ssh/environment
    PATH=PATH:/opt/puppetlabs/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games

  i239onwbhyvo67w.delivery.puppetlabs.net (ubuntu1504-64-1) executed in 0.01 seconds

  i239onwbhyvo67w.delivery.puppetlabs.net (ubuntu1504-64-1) 15:33:57$ grep ^PATH=.*PATH ~/.ssh/environment
    PATH=PATH:/opt/puppetlabs/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games

  i239onwbhyvo67w.delivery.puppetlabs.net (ubuntu1504-64-1) executed in 0.01 seconds

  i239onwbhyvo67w.delivery.puppetlabs.net (ubuntu1504-64-1) 15:33:57$ apt-get update
 Ign http://osmirror.delivery.puppetlabs.net vivid InRelease
    Ign http://builds.puppetlabs.lan vivid InRelease
    Ign http://builds.puppetlabs.lan vivid Release.gpg
    Ign http://osmirror.delivery.puppetlabs.net vivid-updates InRelease
    Get:1 http://builds.puppetlabs.lan vivid Release [1,607 B]
    Ign http://osmirror.delivery.puppetlabs.net vivid-security InRelease
    Get:2 http://osmirror.delivery.puppetlabs.net vivid Release.gpg [933 B]
    Get:3 http://osmirror.delivery.puppetlabs.net vivid-updates Release.gpg [933 B]
    Get:4 http://osmirror.delivery.puppetlabs.net vivid-security Release.gpg [933 B]
    Get:5 http://builds.puppetlabs.lan vivid/PC1 amd64 Packages [954 B]
    Get:6 http://builds.puppetlabs.lan vivid/PC1 i386 Packages [954 B]
    Get:7 http://osmirror.delivery.puppetlabs.net vivid Release [217 kB]
    Ign http://builds.puppetlabs.lan vivid/PC1 Translation-en_US
    Ign http://builds.puppetlabs.lan vivid/PC1 Translation-en
    Get:8 http://osmirror.delivery.puppetlabs.net vivid-updates Release [63.5 kB]
    Get:9 http://osmirror.delivery.puppetlabs.net vivid-security Release [63.5 kB]
    Get:10 http://osmirror.delivery.puppetlabs.net vivid/main Sources [1,077 kB]
    Get:11 http://osmirror.delivery.puppetlabs.net vivid/restricted Sources [7,162 B]
    Get:12 http://osmirror.delivery.puppetlabs.net vivid/universe Sources [7,017 kB]
    Get:13 http://osmirror.delivery.puppetlabs.net vivid/multiverse Sources [174 kB]
    Get:14 http://osmirror.delivery.puppetlabs.net vivid/main amd64 Packages [1,364 kB]
    Get:15 http://osmirror.delivery.puppetlabs.net vivid/restricted amd64 Packages [15.4 kB]
    Get:16 http://osmirror.delivery.puppetlabs.net vivid/universe amd64 Packages [6,485 kB]
    Get:17 http://osmirror.delivery.puppetlabs.net vivid/multiverse amd64 Packages [134 kB]
    Get:18 http://osmirror.delivery.puppetlabs.net vivid/main i386 Packages [1,361 kB]
    Get:19 http://osmirror.delivery.puppetlabs.net vivid/restricted i386 Packages [15.5 kB]
    Get:20 http://osmirror.delivery.puppetlabs.net vivid/universe i386 Packages [6,486 kB]
    Get:21 http://osmirror.delivery.puppetlabs.net vivid/multiverse i386 Packages [134 kB]
    Get:22 http://osmirror.delivery.puppetlabs.net vivid/main Translation-en [793 kB]
    Hit http://osmirror.delivery.puppetlabs.net vivid/multiverse Translation-en
    Hit http://osmirror.delivery.puppetlabs.net vivid/restricted Translation-en
    Get:23 http://osmirror.delivery.puppetlabs.net vivid/universe Translation-en [4,456 kB]
    Get:24 http://osmirror.delivery.puppetlabs.net vivid-updates/main Sources [95.1 kB]
    Get:25 http://osmirror.delivery.puppetlabs.net vivid-updates/restricted Sources [3,687 B]
    Get:26 http://osmirror.delivery.puppetlabs.net vivid-updates/universe Sources [40.8 kB]
    Get:27 http://osmirror.delivery.puppetlabs.net vivid-updates/multiverse Sources [1,957 B]
    Get:28 http://osmirror.delivery.puppetlabs.net vivid-updates/main amd64 Packages [215 kB]
    Get:29 http://osmirror.delivery.puppetlabs.net vivid-updates/restricted amd64 Packages [13.1 kB]
    Get:30 http://osmirror.delivery.puppetlabs.net vivid-updates/universe amd64 Packages [113 kB]
    Get:31 http://osmirror.delivery.puppetlabs.net vivid-updates/multiverse amd64 Packages [5,195 B]
    Get:32 http://osmirror.delivery.puppetlabs.net vivid-updates/main i386 Packages [213 kB]
    Get:33 http://osmirror.delivery.puppetlabs.net vivid-updates/restricted i386 Packages [12.8 kB]
    Get:34 http://osmirror.delivery.puppetlabs.net vivid-updates/universe i386 Packages [114 kB]
    Get:35 http://osmirror.delivery.puppetlabs.net vivid-updates/multiverse i386 Packages [5,394 B]
    Get:36 http://osmirror.delivery.puppetlabs.net vivid-updates/main Translation-en [106 kB]
    Get:37 http://osmirror.delivery.puppetlabs.net vivid-updates/multiverse Translation-en [2,246 B]
    Get:38 http://osmirror.delivery.puppetlabs.net vivid-updates/restricted Translation-en [2,969 B]
    Get:39 http://osmirror.delivery.puppetlabs.net vivid-updates/universe Translation-en [67.0 kB]
    Get:40 http://osmirror.delivery.puppetlabs.net vivid-security/main Sources [46.3 kB]
    Get:41 http://osmirror.delivery.puppetlabs.net vivid-security/restricted Sources [2,792 B]
    Get:42 http://osmirror.delivery.puppetlabs.net vivid-security/universe Sources [18.0 kB]
    Get:43 http://osmirror.delivery.puppetlabs.net vivid-security/multiverse Sources [1,957 B]
    Get:44 http://osmirror.delivery.puppetlabs.net vivid-security/main amd64 Packages [123 kB]
    Get:45 http://osmirror.delivery.puppetlabs.net vivid-security/restricted amd64 Packages [10.4 kB]
    Get:46 http://osmirror.delivery.puppetlabs.net vivid-security/universe amd64 Packages [57.6 kB]
    Get:47 http://osmirror.delivery.puppetlabs.net vivid-security/multiverse amd64 Packages [5,195 B]
    Get:48 http://osmirror.delivery.puppetlabs.net vivid-security/main i386 Packages [122 kB]
    Get:49 http://osmirror.delivery.puppetlabs.net vivid-security/restricted i386 Packages [10.2 kB]
    Get:50 http://osmirror.delivery.puppetlabs.net vivid-security/universe i386 Packages [57.7 kB]
    Get:51 http://osmirror.delivery.puppetlabs.net vivid-security/multiverse i386 Packages [5,394 B]
    Get:52 http://osmirror.delivery.puppetlabs.net vivid-security/main Translation-en [65.8 kB]
    Get:53 http://osmirror.delivery.puppetlabs.net vivid-security/multiverse Translation-en [2,246 B]
    Get:54 http://osmirror.delivery.puppetlabs.net vivid-security/restricted Translation-en [2,607 B]
    Get:55 http://osmirror.delivery.puppetlabs.net vivid-security/universe Translation-en [35.9 kB]
    Fetched 31.5 MB in 10s (3,053 kB/s)
    Reading package lists...

  i239onwbhyvo67w.delivery.puppetlabs.net (ubuntu1504-64-1) executed in 13.02 seconds

  i239onwbhyvo67w.delivery.puppetlabs.net (ubuntu1504-64-1) 15:34:10$ apt-get install --force-yes  -y puppet-agent
    Reading package lists...
    Building dependency tree...
    Reading state information...
    The following NEW packages will be installed:
      puppet-agent
    0 upgraded, 1 newly installed, 0 to remove and 96 not upgraded.
    Need to get 12.4 MB of archives.
    After this operation, 67.6 MB of additional disk space will be used.
    WARNING: The following packages cannot be authenticated!
      puppet-agent
    Get:1 http://builds.puppetlabs.lan/puppet-agent/da2ae6dfcef8ad1c36cbdf63e8c854ef5f06804a/repos/apt/vivid/ vivid/PC1 puppet-agent amd64 1.3.2.193.gda2ae6d-1vivid [12.4 MB]
    debconf: unable to initialize frontend: Dialog
    debconf: (TERM is not set, so the dialog frontend is not usable.)
    debconf: falling back to frontend: Readline
    debconf: unable to initialize frontend: Readline
    debconf: (This frontend requires a controlling tty.)
    debconf: falling back to frontend: Teletype
    dpkg-preconfigure: unable to re-open stdin:
    Fetched 12.4 MB in 0s (47.5 MB/s)
    Selecting previously unselected package puppet-agent.
    (Reading database ...     (Reading database ... 5%(Reading database ... 10%(Reading database ... 15%(Reading database ... 20%(Reading database ... 25%(Reading database ... 30%(Reading database ... 35%(Reading database ... 40%(Reading database ... 45%(Reading database ... 50%(Reading database ... 55%(Reading database ... 60%    (Reading database ... 65%    (Reading database ... 70%    (Reading database ... 75%    (Reading database ... 80%    (Reading database ... 85%    (Reading database ... 90%    (Reading database ... 95%    (Reading database ... 100%(Reading database ...     63975 files and directories currently installed.)
    Preparing to unpack .../puppet-agent_1.3.2.193.gda2ae6d-1vivid_amd64.deb ...
    Unpacking puppet-agent (1.3.2.193.gda2ae6d-1vivid) ...
    Setting up puppet-agent (1.3.2.193.gda2ae6d-1vivid) ...
    Created symlink from /etc/systemd/system/multi-user.target.wants/puppet.service to /lib/systemd/system/puppet.service.
    Created symlink from /etc/systemd/system/multi-user.target.wants/mcollective.service to /lib/systemd/system/mcollective.service.
    Created symlink from /etc/systemd/system/multi-user.target.wants/pxp-agent.service to /lib/systemd/system/pxp-agent.service.

  i239onwbhyvo67w.delivery.puppetlabs.net (ubuntu1504-64-1) executed in 16.96 seconds

i239onwbhyvo67w.delivery.puppetlabs.net (ubuntu1504-64-1) 15:34:27$ puppet --version
  4.3.2
```

joshcooper commented:
All of these platforms passed:

```
centos4-64a centos4-32a
fedora21-64a fedora21-32a
fedora22-64a fedora22-32a
debian6-64a debian6-32a
debian7-64a debian7-32a
debian8-64a debian8-32a
osx109-64a osx1010-64a osx1011-64a
redhat5-64a redhat5-32a
redhat6-64a redhat6-32a
redhat7-64a sles10-64a
sles10-32a sles11-64a
sles11-32a sles12-64a
solaris10-64a solaris11-64a
ubuntu1004-64a ubuntu1004-32a
ubuntu1204-64a ubuntu1204-32a
ubuntu1404-64a ubuntu1404-32a
ubuntu1504-64a ubuntu1504-32a
windows2008r2-64a windows2008r2-6432a
windows2012r2-64a windows2012r2-6432a
```

kevpl commented:
added `apt-get update` call for 15.10, and tested with facter acceptance for a number of ubuntu variants.

joshcooper commented on lib/beaker/dsl/install_utils/foss_utils.rb:
```diff
@@ -882,6 +878,29 @@ def install_puppetlabs_release_repo_on( hosts, repo = nil, opts = options )
+                                  copy_dir )
+          scp_to( host, repo, host.package_config_dir )
+
+          on( host, 'apt-get update' ) if host['platform'] =~ /ubuntu-15\.?10/
```
The issue is that our ubuntu < 15.10 VMs already have ntpdate installed, so when we go to install `puppet-agent`, `Host::Unix::Pkg#apt_needs_update` is still true, so we do `apt-get update` before installing `puppet-agent`. Our ubuntu 15.10 VM does not have ntpdate installed, so we do `apt-get update` prior to installing `ntpdate` early in the beaker run. Then we install the puppet-agent repo config, but by now `apt_needs_update` is false, so we don't `apt-get update` again. I think we should just always run `apt-get update` after installing the puppet_agent_dev_repo like we used to in https://github.com/puppetlabs/beaker/pull/1044/files#diff-b0b73189db06b67008309aabcc69a743L1020.

Sorry for leading us down this path. I didn't understand why apt-get update was running multiple times earlier. I'll also follow up with RE about why the ubuntu VMs are not consistent.

joshcooper replied:
ubuntu wily issue filed as https://tickets.puppetlabs.com/browse/RE-6227

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-beaker.delivery.puppetlabs.net//job/qe_beaker_btc-intn/2044/
Test PASSed.

kevpl commented:
@joshcooper I've updated this to call `apt-get update` for all platforms that support it, like you recommended.  Please let me know if there's anything else you see on this.

joshcooper commented:
Looks good @kevpl, and passes on ubuntu 15.04 and 15.10. Let me run a quick test on all platforms again while jenkins is running.

joshcooper commented:
@kevpl we're seeing a failure on cumulus: 

```
* Install puppet-agent...
  Fetching: http://builds.delivery.puppetlabs.net/puppet-agent/da2ae6dfcef8ad1c36cbdf63e8c854ef5f06804a/repo_configs/deb/pl-puppet-agent-da2ae6dfcef8ad1c36cbdf63e8c854ef5f06804a-.list
    and saving to tmp/repo_configs/cumulus/pl-puppet-agent-da2ae6dfcef8ad1c36cbdf63e8c854ef5f06804a-.list
  OpenURI::HTTPError: 404 Not Found
```

There is a fix for this in puppet#master that needs to be pulled into beaker to retrieve the correct repo_config

justinstoller commented:
wrt to Josh's comment, the full run can be seen linked in this PR: https://github.com/puppetlabs/ci-job-configs/pull/843

joshcooper commented:
AIX, Solaris SPARC, cisco 5 and 7 still need to be validated. @johnduarte and I will verify the first two. @LuvCurves says cisco7 is ready, but we don't have builds yet, and cisco5 is being worked on.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-beaker.delivery.puppetlabs.net//job/qe_beaker_btc-intn/2051/
Test FAILed.

joshcooper commented:
Arista, AIX 5.3 and Solaris 10 & 11 SPARC passed. So we just need the fix for cumulus and need to verify cisco5 and 7.

joshcooper commented:
Cisco7 fails trying to install ntpdate:

```
$ bundle exec rake ci:test:aio TEST_TARGET=ciscoexr7-64 SHA=5e27c223c410a4d41de2716315e8682156278a80 SUITE_VERSION=1.3.2.203.g5e27c22 TESTS=tests/yaml_backend/00-setup.rb
...
lqhyip61xs0ufgp.delivery.puppetlabs.net (ciscoexr7-64-1) 11:54:35$ yum -y  install ntpdate
  Loaded plugins: app_plugin, downloadonly, protect-packages
  There are no enabled repos.
   Run "yum repolist all" to see the repos you have.
   You can enable repos with yum-config-manager --enable <repo>

lqhyip61xs0ufgp.delivery.puppetlabs.net (ciscoexr7-64-1) executed in 11.52 seconds
Exited: 1
Failed: errored in validate
#<Beaker::Host::CommandFailure: Host 'lqhyip61xs0ufgp.delivery.puppetlabs.net' exited with 1 running:
 yum -y  install ntpdate
Last 10 lines of output were:
	Loaded plugins: app_plugin, downloadonly, protect-packages
	There are no enabled repos.
	 Run "yum repolist all" to see the repos you have.
	 You can enable repos with yum-config-manager --enable <repo>>
```

I would be fine adding the cumulus change, merging this PR, and then coming back to resolve the cisco 5 & 7 issues and then releasing the gem. Sound good? @kevpl @johnduarte @LuvCurves

johnduarte commented:
:+1: @joshcooper 

cc: @kevpl  @LuvCurves

LuvCurves commented:
:+1: @joshcooper

LuvCurves commented:
Issue with both Cisco platforms (validate step in Beaker) will be resolved with this ticket: BKR-665

kevpl commented:
cumulus change added

joshcooper commented:
Cumulus passes. It looks like we're not updating from mirrored cumulus repos, I'll file a separate RE ticket about that.

```
* Install puppet-agent...
  Fetching: http://builds.delivery.puppetlabs.net/puppet-agent/5e27c223c410a4d41de2716315e8682156278a80/repo_configs/deb/pl-puppet-agent-5e27c223c410a4d41de2716315e8682156278a80-cumulus.list
    and saving to tmp/repo_configs/cumulus/pl-puppet-agent-5e27c223c410a4d41de2716315e8682156278a80-cumulus.list
  localhost $ scp tmp/repo_configs/cumulus/pl-puppet-agent-5e27c223c410a4d41de2716315e8682156278a80-cumulus.list cumulus25-64-1:/etc/apt/sources.list.d {:ignore => }

  fmn1of8gmv0akqq.delivery.puppetlabs.net (cumulus25-64-1) 13:26:24$ apt-get update
    Ign http://builds.puppetlabs.lan cumulus Release.gpg
    ...
    Hit http://repo.cumulusnetworks.com CumulusLinux-2.5 Release.gpg
   ...
  fmn1of8gmv0akqq.delivery.puppetlabs.net (cumulus25-64-1) 13:26:26$ apt-get install --force-yes  -y puppet-agent
    Reading package lists...
    Building dependency tree...
    Reading state information...
    The following NEW packages will be installed:
      puppet-agent
    0 upgraded, 1 newly installed, 0 to remove and 6 not upgraded.
    Need to get 19.3 MB of archives.
    After this operation, 66.1 MB of additional disk space will be used.
    WARNING: The following packages cannot be authenticated!
      puppet-agent
    Get:1 http://builds.puppetlabs.lan/puppet-agent/5e27c223c410a4d41de2716315e8682156278a80/repos/apt/cumulus/ cumulus/PC1 puppet-agent amd64 1.3.2.203.g5e27c22-1cumulus [19.3 MB]
    debconf: unable to initialize frontend: Readline
    debconf: (This frontend requires a controlling tty.)
    debconf: falling back to frontend: Teletype
    dpkg-preconfigure: unable to re-open stdin:
    Fetched 19.3 MB in 0s (44.0 MB/s)
    Selecting previously unselected package puppet-agent.
    (Reading database ...     26460 files and directories currently installed.)
    Unpacking puppet-agent (from .../puppet-agent_1.3.2.203.g5e27c22-1cumulus_amd64.deb) ...
    Setting up puppet-agent (1.3.2.203.g5e27c22-1cumulus) ...

  fmn1of8gmv0akqq.delivery.puppetlabs.net (cumulus25-64-1) executed in 2.78 seconds
setup/aio/pre-suite/010_Install.rb passed in 4.95 seconds
```

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-beaker.delivery.puppetlabs.net//job/qe_beaker_btc-intn/2052/
Test PASSed.

joshcooper commented:
:+1:
